File: JS/one.js

/*
=====================================================
1. STRING (PRIMITIVO - INMUTABLE - POR VALOR)
=====================================================
- Representa texto
- Inmutable: no se puede modificar
- Cada cambio crea un nuevo string en memoria
*/

let nombre = "Osmar";
let saludo = `Hola ${nombre}`; // Template literal (RECOMENDADO)

console.log(nombre);
console.log(saludo);

// Intento de mutación (NO FUNCIONA)
nombre[0] = "o";
console.log(nombre); // "Osmar"


/*
=====================================================
2. NUMBER (PRIMITIVO - INMUTABLE - IEEE 754)
=====================================================
- NO hay diferencia entre enteros y decimales
- Puede causar errores de precisión
*/

let edad = 30;
let precio = 99.99;

console.log(edad + precio);

// Error clásico de precisión
console.log(0.1 + 0.2); // 0.30000000000000004

// Forma segura de trabajar decimales
let totalSeguro = Number((0.1 + 0.2).toFixed(2));
console.log(totalSeguro);


/*
=====================================================
3. BIGINT (PRIMITIVO - PARA NÚMEROS GIGANTES)
=====================================================
- Se usa cuando number pierde precisión
- Se define con 'n'
*/

let numeroGrande = 123456789012345678901234567890n;
console.log(numeroGrande);

// ❌ NO se puede mezclar con number
// console.log(numeroGrande + 10); // Error

// ✅ Conversión explícita
console.log(numeroGrande + 10n);


/*
=====================================================
4. BOOLEAN (PRIMITIVO - true / false)
=====================================================
- Usado en lógica y condiciones
*/

let esActivo = true;
let esAdmin = false;

if (esActivo) {
  console.log("Usuario activo");
}


/*
=====================================================
5. UNDEFINED (PRIMITIVO - VALOR DEL SISTEMA)
=====================================================
- Variable declarada pero no inicializada
- JS lo asigna automáticamente
*/

let variableSinValor;
console.log(variableSinValor); // undefined


/*
=====================================================
6. NULL (PRIMITIVO - AUSENCIA INTENCIONAL)
=====================================================
- Se asigna manualmente
- Representa "no hay valor"
*/

let usuario = null;
console.log(usuario);

// Bug histórico de JS
console.log(typeof null); // "object" ❌


/*
=====================================================
7. SYMBOL (PRIMITIVO - VALOR ÚNICO)
=====================================================
- Se usa para evitar colisiones de claves
- Muy usado en librerías y frameworks
*/

const ID = Symbol("id");

let usuarioConSymbol = {
  nombre: "Osmar",
  [ID]: 123
};

console.log(usuarioConSymbol);


/*
=====================================================
8. NaN (NOT A NUMBER)
=====================================================
- Es un number inválido
- Nunca es igual a sí mismo
*/

let resultado = Number("hola");
console.log(resultado); // NaN

console.log(resultado === NaN); // false ❌

// Forma correcta de validarlo
console.log(Number.isNaN(resultado)); // true


/*
=====================================================
9. OBJECT (NO PRIMITIVO - MUTABLE - POR REFERENCIA)
=====================================================
- Incluye objetos, arrays, funciones, fechas, etc.
*/

let persona = {
  nombre: "Osmar",
  edad: 30
};

// Mutación directa (PELIGROSA)
persona.edad = 31;
console.log(persona);


/*
=====================================================
10. ARRAY (OBJETO ESPECIAL - MUTABLE)
=====================================================
*/

let numeros = [1, 2, 3];

// Mutación directa
numeros.push(4);
console.log(numeros);

// Forma inmutable (RECOMENDADA)
let nuevosNumeros = [...numeros, 5];
console.log(nuevosNumeros);


/*
=====================================================
11. FUNCTION (OBJETO EJECUTABLE)
=====================================================
- También es un objeto
- Se puede pasar como parámetro
*/

function saludar(nombre) {
  return `Hola ${nombre}`;
}

console.log(saludar("Osmar"));


/*
=====================================================
12. PASO POR VALOR VS REFERENCIA
=====================================================
*/

// POR VALOR (PRIMITIVOS)
let a = 10;
let b = a;

b = 20;
console.log(a); // 10
console.log(b); // 20

// POR REFERENCIA (OBJETOS)
let obj1 = { x: 1 };
let obj2 = obj1;

obj2.x = 99;
console.log(obj1.x); // 99 😱


/*
=====================================================
13. typeof (CON TRAMPAS)
=====================================================
*/

console.log(typeof "hola");       // string
console.log(typeof 123);          // number
console.log(typeof true);         // boolean
console.log(typeof undefined);    // undefined
console.log(typeof null);         // object ❌
console.log(typeof []);           // object ❌
console.log(typeof function(){}); // function

// Validación correcta de arrays
console.log(Array.isArray([])); // true


/*
=====================================================
14. COMPARACIÓN == VS ===
=====================================================
*/

// ❌ Comparación débil (EVITAR)
console.log("5" == 5); // true

// ✅ Comparación estricta (USAR SIEMPRE)
console.log("5" === 5); // false


/*
=====================================================
15. RESUMEN MENTAL (IMPORTANTE)
=====================================================

- Primitivos → inmutables → por valor → stack
- Objetos → mutables → por referencia → heap
- undefined → JS
- null → desarrollador
- === SIEMPRE
- Mutaciones = bugs en proyectos grandes
*/

File: Python/one.py


# =====================================================
# 1. int (ENTEROS - INMUTABLE)
# =====================================================
# - Números enteros
# - Precisión arbitraria (no hay overflow)
# - Inmutable

edad = 30
poblacion = 123456789012345678901234567890

print(edad)
print(poblacion)

# Inmutabilidad
x = 10
y = x
y = 20

print(x)  # 10
print(y)  # 20


# =====================================================
# 2. float (DECIMALES - INMUTABLE)
# =====================================================
# - Usa IEEE 754 (igual que JS)
# - Puede tener errores de precisión

precio = 99.99
total = 0.1 + 0.2

print(precio)
print(total)  # 0.30000000000000004

# Forma segura de trabajar decimales
total_seguro = round(0.1 + 0.2, 2)
print(total_seguro)


# =====================================================
# 3. str (TEXTO - INMUTABLE)
# =====================================================
# - Secuencia de caracteres Unicode
# - Inmutable

nombre = "Osmar"
saludo = f"Hola {nombre}"

print(nombre)
print(saludo)

# Intento de mutación (NO FUNCIONA)
# nombre[0] = "o"  # ❌ TypeError


# =====================================================
# 4. bool (BOOLEANOS)
# =====================================================
# - True / False
# - Subclase de int (True = 1, False = 0)

es_activo = True
es_admin = False

print(es_activo + es_admin)  # 1

# Truthy / Falsy en Python
# Falsy:
# False, 0, 0.0, "", None, [], {}, set()

if "false":
    print("Esto se ejecuta")


# =====================================================
# 5. None (AUSENCIA DE VALOR)
# =====================================================
# - Equivalente a null
# - Representa "nada"

usuario = None
print(usuario)

# Comparación correcta
print(usuario is None)  # True


# =====================================================
# 6. list (MUTABLE)
# =====================================================
# - Colección ordenada
# - Permite duplicados
# - Mutable

numeros = [1, 2, 3]
numeros.append(4)

print(numeros)

# Copia correcta (NO referencia)
nueva_lista = numeros.copy()
nueva_lista.append(5)

print(numeros)
print(nueva_lista)


# =====================================================
# 7. tuple (INMUTABLE)
# =====================================================
# - Colección ordenada
# - Inmutable
# - Más segura que list

coordenadas = (10, 20)
print(coordenadas)


# =====================================================
# 8. dict (MUTABLE)
# =====================================================
# - Pares clave-valor
# - Mutable

persona = {
    "nombre": "Osmar",
    "edad": 30
}

persona["edad"] = 31
print(persona)


# =====================================================
# 9. set (MUTABLE)
# =====================================================
# - Elementos únicos
# - No ordenados

numeros_unicos = {1, 2, 3, 3}
numeros_unicos.add(4)

print(numeros_unicos)


# =====================================================
# 10. PASO POR ASIGNACIÓN DE REFERENCIA
# =====================================================

def modificar_lista(lista):
    lista.append(99)

lista_original = [1, 2, 3]
modificar_lista(lista_original)

print(lista_original)  # [1, 2, 3, 99] 😱


# Forma segura (inmutable)
def modificar_lista_seguro(lista):
    return lista + [99]

lista_nueva = modificar_lista_seguro(lista_original)
print(lista_nueva)


# =====================================================
# 11. type() y isinstance()
# =====================================================

print(type(10))
print(type("hola"))

print(isinstance(10, int))
print(isinstance([], list))


# =====================================================
# 12. RESUMEN MENTAL (CLAVE)
# =====================================================

"""
INMUTABLES:
- int
- float
- bool
- str
- tuple
- None

MUTABLES:
- list
- dict
- set

REGLAS SENIOR:
- Nunca mutar argumentos sin avisar
- Usar copias para listas y dicts
- is None, no ==
- Python NO hace coerción peligrosa
"""
